app/com/arpnetworking/rollups/RollupGenerator.java

```diff
@@ -297,45 +300,62 @@ private LastDataPointMessage buildLastDataPointResponse(
         final String rollupMetricName = message.getMetricName() + message.getPeriod().getSuffix();
         final RollupPeriod period = message.getPeriod();
 
+        // The last datapoint contains data for the period that follows it
         final Instant lastDataPoint = message.getLastDataPointTime().orElse(Instant.EPOCH);
         final Instant oldestBackfillPoint = period.recentEndTime(_clock.instant()).minus(period.periodCountToDuration(_maxBackFillPeriods));
-        if (lastDataPoint.isBefore(oldestBackfillPoint)) {
-            queryBuilder.setStartTime(oldestBackfillPoint);
-        } else {
-            queryBuilder.setStartTime(lastDataPoint);
-        }
+        final Instant startOfRecentClosedPeriod = period.recentStartTime(_clock.instant());
+
+        // We either want to start at the oldest backfill point or the start of the period after the last datapoint.
+        Instant rollupPeriodStart = lastDataPoint.isBefore(oldestBackfillPoint)
+                ? oldestBackfillPoint : lastDataPoint.plus(period.periodCountToDuration(1));
+
+        final Queue<Instant> startTimes = new LinkedList<>();
 
-        // We subtract a millisecond from the endTime because KairosDB treats both start and end time as being
-        // inclusive.  If we don't then we will save a rollup for the next period which will be incomplete.
-        queryBuilder.setEndTime(period.recentEndTime(_clock.instant()).minusMillis(1));
+        // We need to rollup every period up to and including the most recent period.
+        while (rollupPeriodStart.isBefore(startOfRecentClosedPeriod) || rollupPeriodStart.equals(startOfRecentClosedPeriod)) {
+            startTimes.add(rollupPeriodStart);
+            rollupPeriodStart = rollupPeriodStart.plus(period.periodCountToDuration(1));
+        }
 
         metricBuilder.setName(message.getMetricName());
         if (!message.getTags().isEmpty()) {
             metricBuilder.setGroupBy(ImmutableList.of(
                     new MetricsQuery.GroupBy.Builder()
-                    .setName("tag").addOtherArg("tags", message.getTags())
-                    .build()
+                            .setName("tag").addOtherArg("tags", message.getTags())
```

Comment:
nit: can you move `.addOtherArg` to a new line

Reply:
Fixed